### PR TITLE
MNT: update lightpath use for object database loading

### LINKED ISSUE #349: MNT: update usage of lightpath and happi
## Expected Behavior
hutch-python should properly utilize the updated lightpath config.  

## Current Behavior
Currently the hutch-python helpers assume the beamline config contains a list of beamlines that can be searched in happi

https://github.com/pcdshub/hutch-python/blob/ce078b74a6ba13743af177b18b1a31df9120c548/hutch_python/happi.py#L54-L56

`lightpath.config.beamlines` no longer contains a mapping from endstation to "beamlines" contained in the path.  ([gh](https://github.com/pcdshub/lightpath/blob/5d5b36e5c6bfff90414def96e46849a6b58e532d/lightpath/config.py#L7-L8))

`lightpath.config.beamlines` now contains a mapping from endstation to final branch name ([gh](https://github.com/pcdshub/lightpath/blob/aad07bd4817b046e350f794f289830f91e8c7d83/lightpath/config.py#L8))

## Possible Solution
Load the facility graph and grab devices from the active BeamPath.  Let lightpath do the work, graph construction should be fast, and hutch-python already asks for the beam path anyway

## Context
With new lightpath updates hutch-python will fail to load the happi database objects.  This will need to be merged all at once with the lightpath fixes in order to not break hutch-python.

## Your Environment
pcds-5.4.2 + lightpath updates + updated device_config

### PR DESCRIPTION
**Pull request by tangkong**
## Description
- updates the load database step to use new lightpath functionality

## Motivation and Context
Closes #349 

Should only be merged once other lightpath-related PR's are merged (eg pcdsdevices#1028)

Currently seems to cause a slowdown that I have not isolated.  It may be to an increased number of devices being loaded, but testing has not been comprehensive

## How Has This Been Tested?
interactively, with quite a bit of pathmunging

## Where Has This Been Documented?
This PR

### CODE REVIEW

**Comment by ZLLentz** [comment]
Forgot about this PR! Should we wait to tag this repo until lightpath is tagged and available for the CI here?

**Comment by ZLLentz** [comment]
lightpath is already tagged actually, I'm going to close and re-open this to see what the CI says

**Comment by tangkong** [comment]
I totally forgot I did this too.  I'll take another look

**Comment by ZLLentz** [comment]
> the experiments-loading bit is confounding travis

That's annoying but we can ignore it for this PR

**Comment by ZLLentz** [comment]
I think those test failures can be traced back to:
```
16:15:28.690 client          WARNING  MainThread Entry for tst_device_1 is malformed (The information relating to the container class has been modified to the point where the object can not be initialized, please load the corresponding document). Skipping.
16:15:28.706 client          WARNING  MainThread Entry for tst_device_2 is malformed (The information relating to the container class has been modified to the point where the object can not be initialized, please load the corresponding document). Skipping.
```
Which implies that the tst db file needs some additional reworking outside of this PR...

**Comment by ZLLentz** [comment]
Or maybe it means that the tests need happi v2 to pass

**Comment by tangkong** [comment]
No I'm dumb, you're right.  The container needs to not rely on lightpath as well, while still working if lightpath exists.

**Comment by ZLLentz** [comment]
is lightpath not a dependency of hutch-python? I guess it's an optional? We can just force it to be included in the test suite run

**Comment by tangkong** [comment]
I'll give that a shot then

**Comment by tangkong** [comment]
Ah ok at this point we're waiting for pcdsdevices tag to pull the new LCLSLightpathItem container.

**Comment by ZLLentz** [comment]
Gotcha, that should be tagged today but I doubt it will make it through the Friday conda-forge pipeline quickly

**Comment by tangkong** [comment]
With https://github.com/pcdshub/pcdsdevices/pull/1065 I'm reverting the containers to LCLSItem, but hopefully leaving the device_class as a SimpleNamespace so that the branches are accessible after being instantiated. 

Let's see how fast the tag shows up for pypi

**Comment by tangkong** [comment]
This has turned into a test-suite touchup PR.  There's a lightpath bugfix that lets LightController properly handle user supplied configs, but until that is in and tagged some of these tests will be skipped.

This might warrant a quick test run before we merge this, but I've tested loading using a conf.yml that points to an updated happi database

File: hutch_python/tests/test_tstpython.py [hunk 1]
```diff
@@ -3,8 +3,11 @@
 from subprocess import STDOUT, CalledProcessError, check_output
 
 import pytest
+from packaging import version
```
Comment (klauer) [inline hutch_python/tests/test_tstpython.py:6]: `packaging` isn't part of the standard library (https://pypi.org/project/packaging/) and must be a dependency-of-a-dependency if it's not in our requirements files.

Given that it's just in the test suite, we should probably add it to dev-requirements

File: hutch_python/tests/test_tstpython.py [hunk 2]
```diff
@@ -27,16 +30,28 @@ def test_tstpython_scripts():
 
 try:
     import colorama  # noqa
-    has_colorama = True
+
+    # has_colorama = True
+    has_colorama = False
```
Comment (klauer) [inline hutch_python/tests/test_tstpython.py:35]: Was this left in intentionally?

File: hutch_python/tests/test_happi.py [hunk 3]
```diff
@@ -2,11 +2,12 @@
 import os.path
 import tempfile
 
+import pytest
 import simplejson
 
 from hutch_python.happi import get_happi_objs, get_lightpath
 
-from .conftest import beamlines, requires_lightpath
+from .conftest import beamlines, requires_lightpath, sources
```
Comment (klauer) [inline hutch_python/tests/test_happi.py:10]: Not required to change but my opinion is:
* Generic names getting imported from some module are generally not so great
* Instead, importing the module and using `module.generic_name` is preferable

e.g., `conftest.sources`, `conftest.beamlines`

File: hutch_python/happi.py [hunk 4]
```diff
@@ -1,33 +1,47 @@
...
     from lightpath.config import beamlines
 except ImportError:
     lightpath = None
+    LightController = None
     beamlines = None
 
 logger = logging.getLogger(__name__)
 
 
-def get_happi_objs(db, hutch):
```
Comment (klauer) [inline hutch_python/happi.py:17]: (We most likely didn't use this old function signature elsewhere, right? I doubt it...)

The new implementation makes a lot of sense and is a big improvement 👍

File: hutch_python/tests/test_tstpython.py (hunk 2 above)
Comment (tangkong) [reply hutch_python/tests/test_tstpython.py:35]: 🤦  I was clearly not all together when I wrapped this up on Friday

File: hutch_python/tests/test_tstpython.py (hunk 1 above)
Comment (tangkong) [reply hutch_python/tests/test_tstpython.py:6]: I went around looking for a standard library solution to this, but it seems that the similar `distutils.version.LooseVersion` is being deprecated in favor of `packaging.version`.  I'll add this to dev-requirements

**Review by ZLLentz — APPROVED** [review]
I think this looks good and works well in practice. A few of the properties end up being difficult to manage with the pprint nonsense from pcdsdevices but there's not much that can be done about that here.

Example:
```
In [4]: tmo_beampath.blocking_devices
Out[4]:
[VGC(prefix='AT1K0:GAS:VGC:1', name='at1k0_gas_vgc_1', read_attrs=['valve_position', 'open_command', 'interlock_ok', 'open_do', 'error_reset', 'state', 'open_limit', 'closed_limit', 'diff_press_ok', 'ext_ilk_ok', 'at_vac', 'error'], configuration_attrs=['at_vac_setpoint', 'setpoint_hysterisis', 'interlock_device_upstream', 'interlock_device_downstream']),
 VGC(prefix='EM2K0:XGMD:VGC:1', name='em2k0_xgmd_vgc_1', read_attrs=['valve_position', 'open_command', 'interlock_ok', 'open_do', 'error_reset', 'state', 'open_limit', 'closed_limit', 'diff_press_ok', 'ext_ilk_ok', 'at_vac', 'error'], configuration_attrs=['at_vac_setpoint', 'setpoint_hysterisis', 'interlock_device_upstream', 'interlock_device_downstream']),
 VGC(prefix='EM2K0:XGMD:VGC:2', name='em2k0_xgmd_vgc_2', read_attrs=['valve_position', 'open_command', 'interlock_ok', 'open_do', 'error_reset', 'state', 'open_limit', 'closed_limit', 'diff_press_ok', 'ext_ilk_ok', 'at_vac', 'error'], configuration_attrs=['at_vac_setpoint', 'setpoint_hysterisis', 'interlock_device_upstream', 'interlock_device_downstream']),
 VGC(prefix='TV2K0:VGC:1', name='tv2k0_vgc_1', read_attrs=['valve_position', 'open_command', 'interlock_ok', 'open_do', 'error_reset', 'state', 'open_limit', 'closed_limit', 'diff_press_ok', 'ext_ilk_ok', 'at_vac', 'error'], configuration_attrs=['at_vac_setpoint', 'setpoint_hysterisis', 'interlock_device_upstream', 'interlock_device_downstream']),
 VFS(prefix='TV2K0:VFS:1', name='tv2k0_vfs_1', read_attrs=['valve_position', 'vfs_state', 'request_close', 'request_open', 'reset_vacuum_fault', 'override_mode', 'override_force_open', 'gfs_name', 'gfs_trigger', 'position_close', 'position_open', 'vac_fault_ok', 'mps_ok', 'veto_device'], configuration_attrs=[]),

im2k0
-----
units: None

<... snip... hundreds of lines of status printouts ...>
  plc
  ---
    units: None
    status:
    err_code: 0
    cmd_err_reset: 0
    cmd_home: 0,
 VGC(prefix='PC5K4:VWC:01', name='pc5k4_vwc_01', read_attrs=['valve_position', 'open_command', 'interlock_ok', 'open_do', 'error_reset', 'state', 'open_limit', 'closed_limit', 'diff_press_ok', 'ext_ilk_ok', 'at_vac', 'error'], configuration_attrs=['at_vac_setpoint', 'setpoint_hysterisis', 'interlock_device_upstream', 'interlock_device_downstream']),
 TMO Slit sl2k4 on GS2
(hg, vg): (-0.0001, -0.0001) []
(ho, vo): (+15.0000, +15.0000) []
]

In [5]: devs = tmo_beampath.blocking_devices

In [6]: for dev in devs:
   ...:     print(dev.name)
   ...:
at1k0_gas_vgc_1
em2k0_xgmd_vgc_1
em2k0_xgmd_vgc_2
tv2k0_vgc_1
tv2k0_vfs_1
im2k0
mr1k1_vgc_1
mr1k1_bend
tv3k0_vgc_1
mr1k4_vgc_1
mr1k4_soms
tv1k4_vgc_01
tv1k4_vgc_02
st1k4
st3k4_pps
st3k4_pps_vgc_01
mr2k4_kbo_vwc_01
mr2k4_kbo
mr3k4_kbo
mr3k4_kbo_vwc_01
li3k4_ip1_vwc_01
sp1k4
pc5k4_vwc_01
sl2k4

```

File: hutch_python/happi.py [hunk 5]
```diff
@@ -36,32 +50,62 @@ def get_happi_objs(db, hutch):
...
-        blc = [res.item for res in results]
+        blc = [res.item for res in results
+               if res.item.name not in dev_names]
         # Add the beamline containers to the complete list
         if blc:
             containers.extend(blc)
         else:
             logger.warning("No items found in database for %s",
-                           beamline.upper())
+                           line.upper())
```
Comment (ZLLentz) [inline hutch_python/happi.py:93]: One small snag, maybe?
```
WARNING  get_happi_objs - No items found in database for KFE
```

Comment (ZLLentz) [reply hutch_python/happi.py:93]: I think there aren't any missing devices, I'm not sure why we hit this block

Comment (tangkong) [reply hutch_python/happi.py:93]: Good catch.  I think the functionality is fine, but the error message is wrong.  In this case the LightController grabbed a set of devices with `beamline=KFE`, but must have gathered them all, resulting in the additional `blc` list being empty.

Comment (tangkong) [reply hutch_python/happi.py:93]: I'll revise this error message

File: hutch_python/happi.py [hunk 6]
```diff
@@ -36,32 +50,63 @@ def get_happi_objs(db, hutch):
...
+    if light_ctrl is None:
+        # lightpath was unavailable, search by beamline name
+        reqs = dict(beamline=endstation.upper(), active=True)
+        results = client.search(**reqs)
+        containers.extend(res.item for res in results)
+        return _load_devices(*containers)
+
+    # if lightpath exists, we can grab upstream devices
+    dev_names = set()
+    paths = light_ctrl.beamlines[endstation.upper()]
```
Comment (ZLLentz) [inline hutch_python/happi.py:68]: `tstpython`'s error:
```
2022-10-25 15:17:20 - PID 3751      controller.py: 159 load_beamline      WARNING  - Unable to find TST as a configured endstation, assuming this is an invalid path
2022-10-25 15:17:20 - PID 3751           utils.py: 66  safe_load          ERROR    - Failed to load database after 0.19 s
2022-10-25 15:17:20 - PID 3751           utils.py: 67  safe_load          DEBUG    - 'TST'
Traceback (most recent call last):
  File "/cds/home/z/zlentz/pydev/hutch_python/utils.py", line 60, in safe_load
    yield
  File "/cds/home/z/zlentz/pydev/hutch_python/load_conf.py", line 443, in load_conf
    happi_objs = get_happi_objs(db, lc, hutch)
  File "/cds/home/z/zlentz/pydev/hutch_python/happi.py", line 68, in get_happi_objs
    paths = light_ctrl.beamlines[endstation.upper()]
KeyError: 'TST'
```

Comment (ZLLentz) [reply hutch_python/happi.py:68]: Do we need a special case here for when the user asks for an... interesting path?

Comment (ZLLentz) [reply hutch_python/happi.py:68]: Maybe up at line 59 we dip into there for invalid beamline names?

**Comment by ZLLentz** [comment]
I think this is good to go, right?

**Comment by tangkong** [comment]
Thanks for the merge 👍

**Pull request merged**